src/ln/channelmanager.rs

```diff
@@ -1085,11 +1085,22 @@ impl ChainListener for ChannelManager {
 		}
 	}
 
+	/// We don't close the channel, we "halt" it, so no need to send a Shutdown message give us back
+	/// by close_channel to remote peer, we don't want to initiate a closing negotiation
 	fn block_disconnected(&self, header: &BlockHeader) {
-		let mut channel_state = self.channel_state.lock().unwrap();
-		for channel in channel_state.by_id.values_mut() {
-			if channel.block_disconnected(header) {
-				//TODO Close channel here
+		let mut closing_channel = Vec::new();
+		{
+			let mut channel_state = self.channel_state.lock().unwrap();
+			for channel in channel_state.by_id.values_mut() {
+				if channel.block_disconnected(header) {
+					closing_channel.push(channel.get_channel_id().clone());
+				}
+			}
+		}
+		for channel in closing_channel {
+			match self.close_channel(&channel) {
+				Ok(_msg) => {}
+				Err(_e) => {} //TODO: if error is stir up by pending HTLC, after timer, try to call close_channel again ?
```

Comment:
I'm thinking we have 2 way to handle errors (like pending HTLCs) on channel close: 1/ we return a vec of tuple (channel_id, false) to caller of block_disconnected, 2/ we push a timer on event with a new call to block_disconnected. Former is cumbersome for the caller, latter for dealing internally with timer events.

Reply:
I think we need to be more aggressive. Generally, when dealing with a confirmation-tx-unconf error, we'd really prefer to force-close and initiate the race instead of letting our counterparty participate in the shutdown, as its more of a last-ditch thing. Should be able to just do a force_shutdown() and remove the channel from our channels maps, I think.